### PR TITLE
backlight: support to specify percentage of minimum brightness

### LINKED ISSUE #29328: systemd-backlight saves value 0 when notebook lid is closed on service stop
### systemd version the issue has been seen with

systemd-254.3-1-x86_64

### Used distribution

Arch Linux

### Linux kernel version used

6.4.12-arch1-1

### CPU architectures issue was seen on

x86_64

### Component

other

### Expected behaviour you didn't see

When the notebook-lid is closed while attached to an external screen and the service `systemd-backlight@backlight:intel_backlight` get's stopped, the value that's saved in `/sys/class/backlight/intel_backlight/brightness` should get saved into `/var/lib/systemd/backlight/pci-0000:00:02.0:backlight:intel_backlight`.

So when starting the service, the correct value can be loaded.

### Unexpected behaviour you saw

When the lid is closed, instead the value 0 gets saved into `/var/lib/systemd/backlight/pci-0000:00:02.0:backlight:intel_backlight` which ends up in correcting the value to 5% of `max_brightness` on service start.

### Steps to reproduce the problem

To test the behaviour the following steps have been executed:

* Connect to target machine via SSH
* Increase Display Brightness on notebook to at least 50%.
* Check value with `cat /sys/class/backlight/intel_backlight/brightness`
* Close the LID
* Save the value by stopping the service: `systemctl stop systemd-backlight@backlight:intel_backlight`
* Check the saved value with `cat /var/lib/systemd/backlight/pci-0000\:00\:02.0\:backlight\:intel_backlight`
* Open the LID
* Load the saved value by starting the service: `systemctl start systemd-backlight@backlight:intel_backlight`
* Output: `systemd-backlight[17552]: intel_backlight: Saved brightness 0 is too low; increasing to 969.`

### Additional program output to the terminal or log subsystem illustrating the issue

_No response_

### PR DESCRIPTION
**Pull request by yuwata**
Closes #29328.

### CODE REVIEW

File: src/backlight/backlight.c [hunk 1]
```diff
@@ -361,7 +368,7 @@ static int clamp_brightness(sd_device *device, bool saved, unsigned max_brightne
         return 0;
 }
 
-static bool shall_clamp(sd_device *d) {
+static bool shall_clamp(sd_device *d, unsigned *ret) {
         const char *s;
         int r;
```
Comment (YHNdnzj) [inline src/backlight/backlight.c:374]: Missing assertion for `ret`?

**Review by YHNdnzj — APPROVED** [review]

**Comment by yuwata** [comment]
@YHNdnzj Thank you for the review. The missing assertion was added. Upgrading the green label.

File: man/systemd-backlight@.service.xml [hunk 2]
```diff
@@ -31,16 +31,14 @@
...
-    reliably set a brightness value which does not turn off the
-    display.</para>
+    <para><filename>systemd-backlight@.service</filename> is a service that restores the display backlight
+    brightness at early boot and saves it at shutdown. On disk, the backlight brightness is stored in
+    <filename>/var/lib/systemd/backlight/</filename>. During loading, if the udev property
+    <option>ID_BACKLIGHT_CLAMP</option> is not set to false, the brightness is clamped to a value of at least
+    1 or 5% of maximum brightness, whichever is greater. The percentage can be adjusted by specifying a
+    percentage (needs to be suffixed with <literal>%</literal>, e.g. <literal>30%</literal>) to the property
+    <option>ID_BACKLIGHT_CLAMP</option>. This restriction will be removed when the kernel allows user space
+    to reliably set a brightness value which does not turn off the display.</para>
```
Comment (poettering) [inline man/systemd-backlight@.service.xml]: Hmm, so, regarding the last sentence here: first of all, that point in time where we can drop it cannot be determined (i mean, what you are going to do, test all hw and decide all good now?).

Also, why drop the feature, it sounds ok to have generally, as long as we don't expect people to use.

File: src/backlight/backlight.c [hunk 3]
```diff
@@ -361,26 +368,36 @@ static int clamp_brightness(sd_device *device, bool saved, unsigned max_brightne
         return 0;
 }
 
-static bool shall_clamp(sd_device *d) {
+static bool shall_clamp(sd_device *d, unsigned *ret) {
```
Comment (poettering) [inline src/backlight/backlight.c:371]: maybe just make this return the percentage as `int`, and return 0 if we shall not clamp?

(and maybre rename to shall_clamp_percent()?

File: man/systemd-backlight@.service.xml (hunk 2 above)
Comment (poettering) [reply man/systemd-backlight@.service.xml]: (i understand that the sentence was pre-existing, all i am saying is that we should just drop it)

Comment (yuwata) [reply man/systemd-backlight@.service.xml]: OK. Dropped.

File: src/backlight/backlight.c (hunk 3 above)
Comment (yuwata) [reply src/backlight/backlight.c:371]: I initially thought so, but people may want to set `ID_BACKLIGHT_CLAMP=0%`, to set the minimum brightness 1. To support such, we need to return `true` with `*ret = 0`. So, I'd like to keep the code as is.

**Comment by yuwata** [comment]
The documentation is updated. Upgrading the green label.

**Pull request merged**